File: proyecto_java_estatuto/src/main/java/co/edu/udistrital/estatuto/controlador/DocenteControlador.java
package co.edu.udistrital.estatuto.controlador;

import co.edu.udistrital.estatuto.modelo.Docente;
import co.edu.udistrital.estatuto.modelo.TipoVinculacion;
import co.edu.udistrital.estatuto.persistencia.JsonDAO;
import co.edu.udistrital.estatuto.util.Validador;
import com.google.gson.reflect.TypeToken;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * Controlador para la gestión de Docentes.
 *
 * <p>Aplica las validaciones de negocio del dominio docente y
 * delega la persistencia al {@link JsonDAO}. Los docentes se
 * identifican por su número de identificación personal.</p>
 *
 * @author Roberto Albeiro Pava-Díaz
 * @version 1.0
 */
public class DocenteControlador {

    private final JsonDAO<Docente, String> dao;
    private static final String RUTA_DATOS = "datos/docentes.json";

    /**
     * Construye el controlador con la ruta de datos por defecto.
     */
    public DocenteControlador() {
        Type tipo = new TypeToken<List<Docente>>() {}.getType();
        this.dao = new JsonDAO<>(RUTA_DATOS, tipo, Docente::getIdentificacion);
    }

    /**
     * Registra un nuevo docente.
     *
     * @param nombre          nombre completo
     * @param identificacion  número de identificación
     * @param correo          correo institucional
     * @param tipo            tipo de vinculación
     * @param escuelaCodigo   código de la escuela adscrita
     * @param categoria       categoría docente (null para no-planta)
     * @return el Docente registrado
     * @throws IOException              si ocurre un error de persistencia
     * @throws IllegalArgumentException si los datos son inválidos
     */
    public Docente registrar(String nombre, String identificacion, String correo,
                              TipoVinculacion tipo, String escuelaCodigo,
                              String categoria) throws IOException {
        if (!Validador.noEsVacia(nombre)) {
            throw new IllegalArgumentException("El nombre del docente no puede estar vacío.");
        }
        if (!Validador.noEsVacia(identificacion)) {
            throw new IllegalArgumentException("La identificación no puede estar vacía.");
        }
        if (correo != null && !correo.isBlank() && !Validador.esCorreoValido(correo)) {
            throw new IllegalArgumentException("El correo no tiene un formato válido.");
        }

        Docente d = new Docente(
            Validador.normalizarNombre(nombre),
            identificacion.trim(),
            correo != null ? correo.trim() : null,
            tipo,
            escuelaCodigo,
            categoria
        );
        dao.guardar(d);
        return d;
    }

    /**
     * Lista todos los docentes.
     * @return lista de docentes registrados
     */
    public List<Docente> listarTodos() {
        try {
            return dao.buscarTodos();
        } catch (IOException e) {
            System.err.println("Error al listar docentes: " + e.getMessage());
            return new ArrayList<>();
        }
    }

    /**
     * Busca un docente por su número de identificación.
     *
     * @param identificacion número de identificación
     * @return {@link Optional} con el docente
     */
    public Optional<Docente> buscarPorId(String identificacion) {
        try {
            return dao.buscarPorId(identificacion.trim());
        } catch (IOException e) {
            return Optional.empty();
        }
    }

    /**
     * Lista los docentes de una escuela específica.
     *
     * @param escuelaCodigo código de la escuela
     * @return lista de docentes adscritos a esa escuela
     */
    public List<Docente> listarPorEscuela(String escuelaCodigo) {
        try {
            return dao.buscarTodos().stream()
                      .filter(d -> escuelaCodigo.equals(d.getEscuelaCodigo()))
                      .toList();
        } catch (IOException e) {
            return new ArrayList<>();
        }
    }

    /**
     * Lista los docentes de planta.
     * @return lista de docentes con vinculación de planta
     */
    public List<Docente> listarPlanta() {
        try {
            return dao.buscarTodos().stream()
                      .filter(Docente::esPlanta)
                      .toList();
        } catch (IOException e) {
            return new ArrayList<>();
        }
    }

    /**
     * Actualiza los datos de un docente.
     *
     * @param docente docente con datos actualizados
     * @throws IOException si ocurre un error de persistencia
     */
    public void actualizar(Docente docente) throws IOException {
        dao.actualizar(docente);
    }

    /**
     * Elimina un docente por su número de identificación.
     *
     * @param identificacion número de identificación
     * @return {@code true} si se eliminó
     */
    public boolean eliminar(String identificacion) {
        try {
            return dao.eliminar(identificacion.trim());
        } catch (IOException e) {
            System.err.println("Error al eliminar docente: " + e.getMessage());
            return false;
        }
    }
}

File: proyecto_rust_estatuto/src/controlador/docente_controlador.rs
//! # DocenteControlador — capa de control para Docentes
//!
//! Gestiona las operaciones CRUD sobre los `Docente`s del sistema.

use crate::excepcion::errores::EstatutoResult;
use crate::modelo::docente::Docente;
use crate::modelo::enums::TipoVinculacion;
use crate::persistencia::dao::GenericDao;
use crate::persistencia::json_dao::JsonDao;

/// Controlador para la gestión de Docentes.
pub struct DocenteControlador {
    dao: JsonDao<Docente>,
}

impl DocenteControlador {
    /// Crea un nuevo controlador usando el archivo JSON en la ruta dada.
    pub fn nuevo(ruta_datos: &str) -> Self {
        let ruta = format!("{}/docentes.json", ruta_datos);
        DocenteControlador {
            dao: JsonDao::nuevo(ruta, |d: &Docente| d.base.identificacion.clone()),
        }
    }

    /// Lista todos los docentes del sistema.
    pub fn listar_todos(&self) -> EstatutoResult<Vec<Docente>> {
        self.dao.buscar_todos()
    }

    /// Registra un nuevo Docente.
    pub fn registrar(&mut self, docente: Docente) -> EstatutoResult<()> {
        self.dao.guardar(docente)
    }

    /// Busca un Docente por su número de identificación.
    pub fn buscar(&self, identificacion: &str) -> EstatutoResult<Option<Docente>> {
        self.dao.buscar_por_id(&identificacion.to_string())
    }

    /// Actualiza los datos de un Docente existente.
    pub fn actualizar(&mut self, docente: Docente) -> EstatutoResult<()> {
        self.dao.actualizar(docente)
    }

    /// Elimina un Docente del sistema.
    pub fn eliminar(&mut self, identificacion: &str) -> EstatutoResult<bool> {
        self.dao.eliminar(&identificacion.to_string())
    }

    /// Lista docentes filtrados por tipo de vinculación.
    pub fn listar_por_vinculacion(&self, tipo: TipoVinculacion) -> EstatutoResult<Vec<Docente>> {
        let todos = self.dao.buscar_todos()?;
        Ok(todos
            .into_iter()
            .filter(|d| d.tipo_vinculacion == tipo)
            .collect())
    }

    /// Lista docentes de planta.
    pub fn listar_planta(&self) -> EstatutoResult<Vec<Docente>> {
        self.listar_por_vinculacion(TipoVinculacion::Planta)
    }

    /// Lista docentes de una escuela específica.
    pub fn listar_por_escuela(&self, escuela_codigo: &str) -> EstatutoResult<Vec<Docente>> {
        let todos = self.dao.buscar_todos()?;
        let codigo = escuela_codigo.to_uppercase();
        Ok(todos
            .into_iter()
            .filter(|d| d.escuela_codigo == codigo)
            .collect())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::modelo::enums::TipoVinculacion;
    use std::env;

    fn controlador_temporal() -> (DocenteControlador, std::path::PathBuf) {
        let dir = env::temp_dir().join(format!("ctrl_doc_{}", uuid::Uuid::new_v4()));
        std::fs::create_dir_all(&dir).unwrap();
        let ctrl = DocenteControlador::nuevo(dir.to_str().unwrap());
        (ctrl, dir)
    }

    fn docente(id: &str, tipo: TipoVinculacion) -> Docente {
        Docente::nuevo("Nombre", id, None, tipo, "ESC-SIS", "Titular")
    }

    #[test]
    fn registrar_y_buscar_docente() {
        let (mut ctrl, dir) = controlador_temporal();
        ctrl.registrar(docente("CC-100", TipoVinculacion::Planta)).unwrap();
        assert!(ctrl.buscar("CC-100").unwrap().is_some());
        std::fs::remove_dir_all(dir).ok();
    }

    #[test]
    fn listar_planta() {
        let (mut ctrl, dir) = controlador_temporal();
        ctrl.registrar(docente("CC-201", TipoVinculacion::Planta)).unwrap();
        ctrl.registrar(docente("CC-202", TipoVinculacion::HoraCatedra)).unwrap();
        let planta = ctrl.listar_planta().unwrap();
        assert_eq!(planta.len(), 1);
        std::fs::remove_dir_all(dir).ok();
    }
}
